src/main/java/com/xwiki/authentication/saml/xwiki/XWikiUserManager.java

```diff
@@ -0,0 +1,204 @@
+    }
+
+    public DocumentReference getOrCreateUserIfNeeded(XWikiContext context, Saml2XwikiAttributes attributes) throws XWikiException {
+        return new UserManagerWithContextAndAttributes(context, attributes).getOrCreateUserIfNeeded();
+    }
+
+    class UserManagerWithContextAndAttributes {
+        private final XWikiContext context;
+        private final Saml2XwikiAttributes attributes;
+
+        public UserManagerWithContextAndAttributes(XWikiContext context, Saml2XwikiAttributes attributes) {
+            this.context = context;
+            this.attributes = attributes;
+        }
+
+        public DocumentReference getOrCreateUserIfNeeded() throws XWikiException {
+            final User user = getXWikiUsername(attributes.nameID);
+
+            final String wikiId = context.getWikiId();
+            try {
+                // Switch to main wiki to force users to be global users
+                context.setWikiId(context.getMainXWiki());
+
+                if(user.exists())
+                    return syncUserFields(user);
+
+                return createUserWithAttributes(user,attributes);
+            } finally {
+                context.setWikiId(wikiId);
+            }
+        }
+
+        private DocumentReference createUserWithAttributes(User user, Saml2XwikiAttributes attributes) throws XWikiException {
+            LOG.info("Will create new user [{}]", user);
+
+            user.createUserWithAttributes(attributes.xwikiAttributes);
+
+            LOG.info("User [{}] has been successfully created", user);
+
+            return user.getUserReference();
+        }
+
+        private User getXWikiUsername(String nameID)
+                throws XWikiException{
+            final Optional<String> validUserName = findUser(nameID);
+
+            if (validUserName.isPresent()) {
+                LOG.debug("Found XWiki User [{}]", validUserName.get());
+                return new User(nameID, getUserReferenceForName(validUserName.get()), context);
+            }
+            final String generatedUserName = generateValidUserName(nameID);
+            LOG.debug("Generated XWiki User [{}]", generatedUserName);
+
+            return new User(nameID, getUserReferenceForName(generatedUserName), context);
+        }
+
+        private Optional<String> findUser(String nameID) throws XWikiException {
+            final String sql =
+                    "SELECT DISTINCT " +
+                            "   obj.name " +
+                            "FROM " +
+                            "   BaseObject AS obj, StringProperty AS nameidprop " +
+                            "WHERE " +
+                            "   obj.className = ?1 " +
+                            "AND " +
+                            "   obj.id = nameidprop.id.id " +
+                            "AND " +
+                            "   nameidprop.id.name = ?2 " +
+                            "AND " +
+                            "   nameidprop.value = ?3";

```

Comment:
We think you maaaaaybe went a bit overboard on the like breaks here
```suggestion
            final String sql =
                    "SELECT DISTINCT obj.name " +
                    "FROM BaseObject AS obj, StringProperty AS nameidprop " +
                    "WHERE obj.className = ?1 AND obj.id = nameidprop.id.id AND nameidprop.id.name = ?2 AND nameidprop.value = ?3";
```